index.d.ts

```diff
@@ -23,12 +24,321 @@ type InitOptions = {
   publishableKey: string;
 };
 
+type OpenCheckoutOptions = {
+  hostedPage: Function;
+  layout: Layout;
+  loaded: Function;
+  error: Function;
+  success(hostedPageId: string): void;
+  close: Function;
+  step(currentStep: string): void;
+};
+
+interface AddressDetails {
+  first_name: string;
+  last_name: string;
+  company: string;
+  phone: string;
+  email: string;
+  line1: string;
+  line2: string;
+  city: string;
+  state: string;
+  state_code: string;
+  country: string;
+  zip: string;
+}
+
+type BillingAddress = AddressDetails;
+
+type ShippingAddress = AddressDetails;
+
+interface CustomerDetails {
+  first_name: string;
+  last_name: string;
+  company: string;
+  phone: string;
+  email: string;
+  billingAddress: BillingAddress;
+}
+
+interface AddOn {
+  id: string;
+  quantity?: number;
+  quantity_in_decimal?: string;
+}
+
+type EventBasedAddon = {
+  id?: string;
+  quantity?: number;
+  unit_price?: number;
+  service_period_in_days?: number;
+  on_event?:
+    | 'subscription_creation'
+    | 'subscription_trial_start'
+    | 'plan_activation'
+    | 'subscription_activation';
+  charge_once?: boolean;
+  charge_on?: 'immediately' | 'on_event';
+};
+
+interface ChargeItem extends AddOn {}
+
+interface Product {
+  setPlanQuantity(planQuantity: number): Product;
+  setPlanQuantityInDecimal(decimalQuantity: string): Product;
+  incrementPlanQuantity(): Product;
+  decrementPlanQuantity(): Product;
+  addAddon(addon: AddOn | string): Product;
+  removeAddon(addon: AddOn | string): Product;
+  setAddons(addons: Array<AddOn>): Product;
+  incrementAddonQty(addonId: string): Product;
+  decrementAddonQty(addonId: string): Product;
+  addCoupon(couponCode: string): Product;
+  removeCoupon(couponCode: string): Product;
+  setCustomData(
+    data: object /** Object containing subs custom fields */
+  ): Product;
+  getLayout(): Layout;
+  setLayout(layout: Layout): void;
+  setCharges(charges: Array<ChargeItem>): Product;
+  addCharge(chargeItem: ChargeItem): Product;
+  removeCharge(chargeItemId: string): Product;
+  incrementChargeQty(chargeItemId: string): Product;
+  decrementChargeQty(chargeItemId: string): Product;
+}
+
+interface Cart {
+  replaceProduct(product: Product): Cart;
+  proceedToCheckout(): void;
+  setBusinessEntity(businessEntityId: string): Promise<void>;
+  setAffiliateToken(token: string): Cart;
+  setCustomer(customer: Customer): Cart;
+  setShippingAddress(shippingAddress: ShippingAddress): Cart;
+}
+
+interface PortalOpenOptions {
```

Comment:
Should this directly be named PortalCallbacks

Reply:
Done.